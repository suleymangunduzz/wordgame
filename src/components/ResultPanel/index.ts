import ResultPanel from './ResultPanel';

export default ResultPanel;
