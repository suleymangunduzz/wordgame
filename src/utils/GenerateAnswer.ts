const generateAnswer = (selectedNames: string[], names: string[]) => {
  const lastValidAnswer = selectedNames[selectedNames.length - 1];

  const lastValidAnswerEndsWith = lastValidAnswer.charAt(lastValidAnswer.length - 1);

  const unusedNameList = names.filter(name => !selectedNames.includes(name));

  const newAnswer = unusedNameList.find(name => name.charAt(0) === lastValidAnswerEndsWith);

  return newAnswer || '';
};

export default generateAnswer;
