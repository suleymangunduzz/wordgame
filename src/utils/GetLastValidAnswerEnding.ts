const getLastValidAnswerEnding = (selectedNames: string[]) => {
  const lastValidAnswer = selectedNames[selectedNames.length - 1];

  return lastValidAnswer.charAt(lastValidAnswer.length - 1);
};

export default getLastValidAnswerEnding;
