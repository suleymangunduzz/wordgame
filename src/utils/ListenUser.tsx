import checkAnswer from './CheckAnswer';

export const listenUser = (
  recognition: any,
  selectedNames: string[],
  names: string[],
  setWinner: (name: string) => void,
  setTimeLeft: (timeLeft: number) => void,
  setGameStatus: (status: string) => void,
  setNextPlayer: (nextPlayer: string) => void,
  setUsersFinalAnswer: (answer: string) => void,
  setSelectedNames: (selectedNames: string[]) => void,
) => {
  console.log("Well, it is user's turn, lets play it !");

  recognition.start();

  // eslint-disable-next-line no-param-reassign
  recognition.onresult = (event: any) => {
    let finalTranscript = '';

    for (let i = event.resultIndex; i < event.results.length; i += 1) {
      const { transcript } = event.results[i][0];
      if (event.results[i].isFinal) finalTranscript += `${transcript} `;
    }
    if (event.results[0].isFinal) {
      console.log("User's answer: ", finalTranscript.toLocaleLowerCase('tr'));
      setUsersFinalAnswer(finalTranscript.toLocaleLowerCase('tr'));
      checkAnswer(
        finalTranscript.toLocaleLowerCase('tr'),
        selectedNames,
        names,
        setGameStatus,
        setNextPlayer,
        setWinner,
        setSelectedNames,
        setTimeLeft,
      );
    }
  };
};

export default listenUser;
