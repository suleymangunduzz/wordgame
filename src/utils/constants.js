export const FINISHED = 'FINISHED';
export const NOT_STARTED = 'NOT_STARTED';
export const STARTED = 'STARTED';

export const USER = 'USER';
export const MACHINE = 'MACHINE';
