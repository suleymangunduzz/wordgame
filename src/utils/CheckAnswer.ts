import getLastValidAnswerEnding from './GetLastValidAnswerEnding';
import { MACHINE, FINISHED } from './constants';

export const checkAnswer = (
  answer: string,
  selectedNames: string[],
  names: string[],
  setGameStatus: (gameStatus: string) => void,
  setNextPlayer: (nextPlayer: string) => void,
  setWinner: (winner: string) => void,
  setSelectedNames: (selectedNames: string[]) => void,
  setTimeLeft: (timeLeft: number) => any,
) => {
  const cleanAnswer = answer.replace(/\s+/g, '');
  const answerStartsWith = cleanAnswer.charAt(0);

  const lastValidAnswerEndsWith = getLastValidAnswerEnding(selectedNames);

  if (selectedNames.some(name => name === cleanAnswer)) {
    console.log('The answer is used before, so the game is finishing.');
    setGameStatus(FINISHED);
    setNextPlayer(MACHINE);
    setWinner(MACHINE);
    return;
  }

  if (answerStartsWith !== lastValidAnswerEndsWith) {
    console.log('The answer does not match with the last valid answer.');
    setGameStatus(FINISHED);
    setNextPlayer(MACHINE);
    setWinner(MACHINE);
    return;
  }

  if (!names.includes(cleanAnswer)) {
    console.log('There is no name like this in our data.');
    setGameStatus(FINISHED);
    setNextPlayer(MACHINE);
    setWinner(MACHINE);
    return;
  }

  console.log('The answer is correct and added to the our list, good job, keep playing!');
  const newState: string[] = [...selectedNames, cleanAnswer];
  setSelectedNames(newState);
  setNextPlayer(MACHINE);
  setTimeLeft(8);
};

export default checkAnswer;
