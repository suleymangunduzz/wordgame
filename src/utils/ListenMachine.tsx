import generateAnswer from './GenerateAnswer';
import { USER, FINISHED } from './constants';

export const listenMachine = (
  isLucky: boolean,
  timeToWait: number,
  selectedNames: string[],
  names: string[],
  synthRef: any,
  setWinner: (name: string) => void,
  setTimeLeft: (timeLeft: number) => void,
  setGameStatus: (status: string) => void,
  setNextPlayer: (nextPlayer: string) => void,
  setUsersFinalAnswer: (answer: string) => void,
  setSelectedNames: (selectedNames: string[]) => void,
  isFirtTime?: boolean,
) => {
  console.log('Machine is playing...');
  setUsersFinalAnswer('');

  if (isLucky) {
    console.log('Luckily, it keeps playing...');

    console.log(`After ${timeToWait} seconds machine is gonna answer.`);

    let newSelection = '';
    let newState: string[] = [];

    if (isFirtTime) {
      newSelection = names[timeToWait];
      newState = [newSelection];
    } else {
      console.log('Generate some valid name for the machine if it is possible.');
      newSelection = generateAnswer(selectedNames, names);
      newState = [...selectedNames, newSelection];
    }

    let utter: any;

    if ('SpeechSynthesisUtterance' in window) {
      utter = new SpeechSynthesisUtterance(newSelection);
      utter.lang = 'tr-TR';
    }

    setTimeout(() => {
      if (newSelection !== '') {
        setSelectedNames(newState);
        setNextPlayer(USER);
        if (synthRef.current) {
          synthRef.current.speak(utter);
        }
        setTimeLeft(8);
      } else {
        setGameStatus(FINISHED);
        setWinner(USER);
      }
    }, timeToWait * 1000);

    clearTimeout();
  } else {
    console.log('Machine can not find a name.');
    console.log('Game is gonna finish, winner is the user.');

    setTimeout(() => {
      setGameStatus(FINISHED);
      setWinner(USER);
    }, 8000);
  }
};

export default listenMachine;
