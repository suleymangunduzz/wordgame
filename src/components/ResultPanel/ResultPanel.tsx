import React, { FC } from 'react';

import './ResultPanel.scss';

import { FINISHED } from '../../utils/constants';

type Props = {
  gameStatus: string;
  selectedNames: Array<string>;
};

const ResultPanel: FC<Props> = ({ gameStatus, selectedNames }) => {
  if (gameStatus === FINISHED) {
    return <div className="ResultPanel">Oyun Bitti</div>;
  }

  return <div>{selectedNames.length ? selectedNames[selectedNames.length - 1] : '......'}</div>;
};

export default ResultPanel;
