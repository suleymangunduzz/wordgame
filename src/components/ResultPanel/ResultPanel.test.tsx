import React from 'react';
import { shallow } from 'enzyme';
import { FINISHED, STARTED } from '../../utils/constants';
import ResultPanel from './ResultPanel';

it('renders correctly Result Panel when game is finished', () => {
  const mockSelectedNames: string[] = [];
  const wrapper = shallow(<ResultPanel gameStatus={FINISHED} selectedNames={mockSelectedNames} />);

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly Result Panel when game is playing but there is no name', () => {
  const mockSelectedNames: string[] = [];
  const wrapper = shallow(<ResultPanel gameStatus={STARTED} selectedNames={mockSelectedNames} />);

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly Result Panel when game is finihed with names', () => {
  const mockSelectedNames: string[] = ['ali', 'ismet', 'temel', 'lale'];
  const wrapper = shallow(<ResultPanel gameStatus={STARTED} selectedNames={mockSelectedNames} />);

  expect(wrapper).toMatchSnapshot();
});
