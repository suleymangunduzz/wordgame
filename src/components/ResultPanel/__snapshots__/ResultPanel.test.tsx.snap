// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly Result Panel when game is finihed with names 1`] = `
<div>
  lale
</div>
`;

exports[`renders correctly Result Panel when game is finished 1`] = `
<div
  className="ResultPanel"
>
  Oyun Bitti
</div>
`;

exports[`renders correctly Result Panel when game is playing but there is no name 1`] = `
<div>
  ......
</div>
`;
